File: src/main/java/com/chtholly/chthl/node/DocumentNode.java
package com.chtholly.chthl.node;

import com.chtholly.chthl.generator.Visitor;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * The root node of a CHTL document's AST. It contains all top-level statements/nodes.
 */
public class DocumentNode implements Node {
    private final List<Node> children = new ArrayList<>();

    public List<Node> getChildren() {
        return children;
    }

    public void addChild(Node node) {
        this.children.add(node);
    }

    @Override
    public String tokenLiteral() {
        if (children.isEmpty()) {
            return "";
        }
        return children.get(0).tokenLiteral();
    }

    @Override
    public String toString() {
        return children.stream()
                .map(Node::toString)
                .collect(Collectors.joining("\n"));
    }

    @Override
    public <T> T accept(Visitor<T> visitor) {
        return visitor.visit(this);
    }
}

File: CHTL/CHTLNode/DocumentNode.cpp
#include "DocumentNode.h"
#include "Visitor.h"

namespace CHTL {

void DocumentNode::accept(Visitor& visitor) {
    visitor.visit(*this);
}

void DocumentNode::addChild(std::unique_ptr<BaseNode> child) {
    children.push_back(std::move(child));
}

std::unique_ptr<BaseNode> DocumentNode::clone() const {
    auto new_node = std::make_unique<DocumentNode>();
    for (const auto& child : this->children) {
        new_node->addChild(child->clone());
    }
    return new_node;
}

} // namespace CHTL
